### PR TITLE
Implement Extend for ArrayBuilder (#1841)

### LINKED ISSUE #1841: Implement Extend for Builder
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The [FromIterator](https://doc.rust-lang.org/std/iter/trait.FromIterator.html) trait is a standard trait in the ecosystem for building a collection from an iterator. Supporting it within the builders would allow writing code like (not tested)

```
let strings = ["1", "2", "3"];

let array = strings
    .iter()
    .map(|s| Some(s.parse()?))
    .collect::<Result<Int32Builder>>()
    .unwrap()
    .finish();
```

**Describe the solution you'd like**

Implement FromIterator for the Builder types. In most cases FromIterator is implemented for the corresponding Array type, and so it should largely just be a case of moving this logic into the builder and making the array implementation call it.

**Describe alternatives you've considered**

We could not do this

### LINKED ISSUE #3562: Panic on Key Overflow in Dictionary Builders
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I suspect this will be a controversial issue, but there a couple of reasons for proposing this

* Rust collections and accompanying traits like `FromIterator` and `Extend` assume infallibility
* We currently panic on offset overflow for StringArray, ByteArray, ListArray, etc...
* We already panic on offset overflow in FromIterator for DictionaryArray
* The error is just going to be surfaced to the user anyway, a panic is perfectly acceptable for this imo

**Describe the solution you'd like**

Panic on key overflow instead of returning an error

**Describe alternatives you've considered**

We could implement `Extend` and panic on key overflow there, much like we do for FromIterator, but I feel we should be consistent about this.

We could also implement our own `TryExtend` trait or something similar.

**Additional context**

#2725 tracks a more holistic rework of how we handle errors

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1841
Closes #3562

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/builder/generic_bytes_dictionary_builder.rs [hunk 1]
```diff
@@ -255,12 +255,34 @@ where
         Ok(key)
     }
 
+    /// Infallibly append a value to this builder
+    ///
+    /// # Panics
+    ///
+    /// Panics if the resulting length of the dictionary values array would exceed `T::Native::MAX`
+    pub fn append_value(&mut self, value: impl AsRef<T::Native>) {
```
Comment (tustvold) [inline arrow-array/src/builder/generic_bytes_dictionary_builder.rs:263]: These are the new APIs for #3562

**Comment by tustvold** [comment]
> What's the benefit of this over including an append_all function?

https://github.com/apache/arrow-rs/pull/3563/files#diff-cb5b791e20e4536940eecb1466e034510c245d0d443fb89942b8ab94171693c8R192 is an example of a use-case that isn't possible without some sort of statically typed trait abstraction, using the standard ecosystem primitive seems better than rolling our own.

File: arrow-array/src/array/dictionary_array.rs [hunk 2]
```diff
@@ -481,18 +481,7 @@ impl<'a, T: ArrowDictionaryKeyType> FromIterator<Option<&'a str>> for Dictionary
...
-                // error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
-                builder
-                    .append(i)
-                    .expect("Unable to append a value to a dictionary array.");
-            } else {
-                builder.append_null();
-            }
-        });
-
+        builder.extend(it);
```
Comment (alamb) [inline arrow-array/src/array/dictionary_array.rs:484]: that is certainly nicer

File: arrow-array/src/builder/generic_list_builder.rs [hunk 3]
```diff
@@ -364,4 +390,25 @@ mod tests {
             list_array.values().data().child_data()[0].buffers()[0].clone()
         );
     }
+
+    #[test]
+    fn test_extend() {
```
Comment (alamb) [inline arrow-array/src/builder/generic_list_builder.rs:395]: 👍  thank you for this test

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @tustvold 

The only thing that might be worth doing is some sort of test coverage (as you have done for ListArray) so that `extend` can't be removed by accident without also changing tests (aka so we don't break them by accident)

**Pull request merged**